### PR TITLE
Resolve data_files by split name

### PR DESCRIPTION
**Pull request by lhoestq**
As discussed in https://github.com/huggingface/datasets/issues/3027 we should automatically infer what file is supposed to go to what split automatically, based on filenames.

I added the support for different kinds of patterns, for both dataset repositories and local directories:

```
    Input structure:

        my_dataset_repository/
        ├── README.md
        └── dataset.csv

    Output patterns:

        {"train": ["*"]}
```
```
    Input structure:

        my_dataset_repository/
        ├── README.md
        ├── train.csv
        └── test.csv

        my_dataset_repository/
        ├── README.md
        └── data/
            ├── train.csv
            └── test.csv

        my_dataset_repository/
        ├── README.md
        ├── train_0.csv
        ├── train_1.csv
        ├── train_2.csv
        ├── train_3.csv
        ├── test_0.csv
        └── test_1.csv

    Output patterns:

        {"train": ["*train*"], "test": ["*test*"]}
```
```
    Input structure:

        my_dataset_repository/
        ├── README.md
        └── data/
            ├── train/
            │   ├── shard_0.csv
            │   ├── shard_1.csv
            │   ├── shard_2.csv
            │   └── shard_3.csv
            └── test/
                ├── shard_0.csv
                └── shard_1.csv

    Output patterns:

        {"train": ["*train*/*", "*train*/**/*"], "test": ["*test*/*", "*test*/**/*"]}
```

and also this pattern that allows to have custom split names, and that is the structure used by #3098 for `push_to_hub` (cc @LysandreJik ):

```
    Input structure:

        my_dataset_repository/
        ├── README.md
        └── data/
            ├── train-00000-of-00003.csv
            ├── train-00001-of-00003.csv
            ├── train-00002-of-00003.csv
            ├── test-00000-of-00001.csv
            ├── random-00000-of-00003.csv
            ├── random-00001-of-00003.csv
            └── random-00002-of-00003.csv

    Output patterns:

        {
            "train": ["data/train-[0-9][0-9][0-9][0-9][0-9]-of-[0-9][0-9][0-9][0-9][0-9].*"],
            "test": ["data/test-[0-9][0-9][0-9][0-9][0-9]-of-[0-9][0-9][0-9][0-9][0-9].*"],
            "random": ["data/random-[0-9][0-9][0-9][0-9][0-9]-of-[0-9][0-9][0-9][0-9][0-9].*"],
        }
```

You can check the documentation about structuring your repository [here](https://52640-250213286-gh.circle-artifacts.com/0/docs/_build/html/repository_structure.html). cc @stevhliu 

Fix https://github.com/huggingface/datasets/issues/3027
Fix https://github.com/huggingface/datasets/issues/3212

In the future we can also add support for dataset configurations.

### CODE REVIEW

**Comment by borisdayma** [comment]
Really cool!
When splitting by folder, what do we use for validation set ("valid", "validation" or both)?

**Comment by lhoestq** [comment]
> When splitting by folder, what do we use for validation set ("valid", "validation" or both)?

Both are fine :) As soon as it has "valid" in it

File: docs/source/repository_structure.rst [hunk 1]
```diff
@@ -0,0 +1,141 @@
+Structure your repository
+=========================
+
+To host and share your dataset, you can create a dataset repository on the Hugging Face Dataset Hub and upload your data files.
+
+This guide aims to explain what structure is expected when you upload a dataset.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:6]:
```suggestion
This guide will show you how to structure your dataset repository when you upload it.
``````

File: docs/source/repository_structure.rst [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+A dataset with a supported structure can be loaded automatically using ``load_dataset``, and also has a nice preview on its dataset page on the Hub.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:7]:
```suggestion
A dataset with a supported structure can be loaded automatically with ``load_dataset``, and it will have a preview on its dataset page on the Hub.
```

File: docs/source/repository_structure.rst [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+.. Note::
+
+    In the following examples, CSV files are used. However you can use any supported format (text, JSON, JSON Lines, CSV, Parquet).
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:11]:
```suggestion
    The following examples use CSV files, but you can use any supported format (text, JSON, JSON Lines, CSV, Parquet).
```

File: docs/source/repository_structure.rst [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+Main use-case
+-------------
+
+Here is an easy structure that you can use for your repository. It contains two files: `train.csv` and `test.csv`.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:16]:
```suggestion
The simplest dataset structure two files: `train.csv` and `test.csv`.
```

File: docs/source/repository_structure.rst [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+It also contains the dataset card `README.md` that is the `dataset card <dataset_card.html>`__ that is displayed on the dataset webpage.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:18]:
```suggestion
Your repository will also contain a `README.md` file, the `dataset card <dataset_card.html>`__ displayed on your dataset page.
```

File: docs/source/repository_structure.rst [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    my_dataset_repository/
+    ├── README.md
+    ├── train.csv
+    └── test.csv
+
+
+Splits and file names
+---------------------
+
+The train/validation/test splits of your dataset are automatically infered from the file names.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:31]:
```suggestion
🤗 Datasets automatically infers the train/validation/test splits of your dataset from the file names. All the files that contain **train** in their names are considered part of the train split. The same idea applies to the test and validation split: 
* All the files that contain **test** in their names are considered part of the test split.
* All the files that contain **valid** in their names are considered part of the validation split.
```

File: docs/source/repository_structure.rst [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    └── data/
+        ├── train.csv
+        ├── test.csv
+        └── valid.csv
+
+
+Multiple files per split
+------------------------
+
+If your train is split into several files it also works.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:49]:
```suggestion
If one of your splits comprises several files, 🤗 Datasets can still infer whether it is the train/validation/ test split from the file name. For example, if your **train** split spans several files:
```

File: docs/source/repository_structure.rst [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    my_dataset_repository/
+    ├── README.md
+    ├── train_0.csv
+    ├── train_1.csv
+    ├── train_2.csv
+    ├── train_3.csv
+    ├── test_0.csv
+    └── test_1.csv
+
+Just make sure that all the files of your train set have "train" inside their names (same for test and validation).
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:62]:
```suggestion
Just make sure that all the files of your **train** set have **train** in their names (same for test and validation). It doesn’t matter if you add a prefix or suffix to **train** in the file name (like `my_train_file_00001.csv`, for example). 🤗 Datasets can still infer the appropriate split.
```

File: docs/source/repository_structure.rst [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+It doesn't matter if you add a prefix before "train" or a suffix after "train" in the file name (like "my_train_file_00001.csv" for example).
+
+For convenience you can also place your data files into different directories. In this case the split name is infered from the directory name.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:65]:
```suggestion
For convenience, you can also place your data files into different directories. In this case, the split name is inferred from the directory name.
```

File: docs/source/repository_structure.rst [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        │   └── shard_3.csv
+        └── test/
+            ├── shard_0.csv
+            └── shard_1.csv
+
+
+Custom split names
+------------------
+
+If you have other splits in addition to the traditionnal train/validation/test, you must use the following structure.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:85]:
```suggestion
If you have other data files in addition to the traditional train/validation/test sets, you must use the following structure. Follow the file name format exactly for this type of structure: `data/<split_name>-xxxxx-of-xxxxx.csv`. Here is an example with three splits: **train**, **test**, and **random**:
```

File: docs/source/repository_structure.rst [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        ├── test-00000-of-00001.csv
+        ├── random-00000-of-00003.csv
+        ├── random-00001-of-00003.csv
+        └── random-00002-of-00003.csv
+
+
+Multiple configuration (WIP)
+----------------------------
+
+You can separate different configurations of your dataset (for example if it is splti in different languages) by using one directory per configuration.
```
Comment (stevhliu) [inline docs/source/repository_structure.rst:106]:
```suggestion
You can specify different configurations of your dataset (for example, if a dataset contains different languages) with one directory per configuration.
```

**Comment by lhoestq** [comment]
Merging for now, if you have comments about the documentation we can address them in subsequent PRs :)

**Review by stevhliu — APPROVED** [review]
Minor suggestions, but everything else looks great! 🎉

I wonder whether we should include the Multiple configurations section right now since it isn't supported yet, and is unlikely to help readers currently. But I can see the value of having it in case users are curious in whether something like this will be supported in the future.

**Comment by lhoestq** [comment]
Thanks for the comments @stevhliu :) I just opened https://github.com/huggingface/datasets/pull/3233 to take them into account
